master/buildbot/steps/artifact.py

```diff
@@ -379,6 +448,108 @@ def start(self):
+        br = yield self.master.db.buildrequests.getBuildRequestTriggered(triggeredbybrid, self.artifactBuilderName)
+        defer.returnValue(br)
+
+
+class DownloadArtifactFromParent(DownloadArtifact):
+    @defer.inlineCallbacks
+    def _getBuildRequest(self):
+        triggeredbybrid = self.build.requests[0].id
+        id = yield self.master.db.buildrequests.getTriggeredById(triggeredbybrid)
+        br = yield self.master.db.buildrequests.getBuildRequestById(id)
+        defer.returnValue(br)
+
+class DownloadArtifactFromChilden(LoggingBuildStep, CompositeStepMixin):
+    def __init__(self, name=None, projectPrefix=None, targetConfig=None, artifactBuilderName=None,
+                 artifact=None, artifactDirectory=None, artifactDestination=None,
+                 artifactServer=None, artifactServerDir=None, artifactServerPort=None,
+                 usePowerShell=True, baseLocalDir=None, **kwargs):
+        self.workdir = 'build'
+        self.artifactBuilderName = artifactBuilderName
+        self.artifact = artifact
+        self.projectPrefix = projectPrefix
+        self.target_config = targetConfig
+        self.artifactDirectory = artifactDirectory
+        self.artifactServer = artifactServer
+        self.artifactServerDir = artifactServerDir
+        self.artifactServerPort = artifactServerPort
+        self.artifactDestination = artifactDestination or artifact
+        self.master = None
+        self.usePowerShell = usePowerShell
+        self.kwargs = kwargs
+        self.name = name
+        self.baseLocalDir = baseLocalDir
+        LoggingBuildStep.__init__(self, **kwargs)
+
+    @defer.inlineCallbacks
+    def start(self):
+        if self.master is None:
+            self.master = self.build.builder.botmaster.parent
+
+        brid = self.build.requests[0].id
+        triggeredBuilderName = self.projectPrefix + self.target_config
+        partitionRequests = yield self.master.db.buildrequests.getBuildRequestsTriggeredBy(brid, triggeredBuilderName)
+        buildRequetsIdsWithArtifacts = self._getBuildRequestIdsWithArtifacts(partitionRequests)
+        self.partitionCount = len(buildRequetsIdsWithArtifacts)
+        self.stdio_log = self.addLogForRemoteCommands("stdio")
+        self.stdio_log.setTimestampsMode(self.timestamp_stdio)
+
+        for id in buildRequetsIdsWithArtifacts:
+            buildRequest = yield self.master.db.buildrequests.getBuildRequestById(id)
+            triggeredBuilderName = self.projectPrefix + self.target_config
+            artifactPath = "%s_%s_%s" % (safeTranslate(triggeredBuilderName),
+                                         id, FormatDatetime(buildRequest["submitted_at"]))
+
+            artifactPath += "/%s" % self.artifactDirectory
+            remotelocation = getRemoteLocation(self.artifactServer, self.artifactServerDir, artifactPath, "")
+            localdir = self.baseLocalDir + '/' + str(id)
+
+            command = [mkDir(self), '-p', '\''+ localdir + '\'']
```

Comment:
not sure if it is this command but the destination directory is really strange with duplicate of each directory, here is the `ls -la` of the build directory:

```drwx------ 8 vagrant vagrant 4096 Jul  3 14:29 ./
drwx------ 3 vagrant vagrant 4096 Jul  3 14:29 ../
drwxrwxr-x 2 vagrant vagrant 4096 Jul  3 14:29 845/
drwx------ 2 vagrant vagrant 4096 Jul  3 14:29 '845'/
drwxrwxr-x 2 vagrant vagrant 4096 Jul  3 14:29 846/
drwx------ 2 vagrant vagrant 4096 Jul  3 14:29 '846'/
drwxrwxr-x 2 vagrant vagrant 4096 Jul  3 14:29 848/
drwx------ 2 vagrant vagrant 4096 Jul  3 14:29 '848'/
```